### PR TITLE
feat: identify exotic co-factors

### PR DESCRIPTION
**Pull request by Midnighter**
Add a celery task to identify non-native co-factors for each predicted pathway found.

### CODE REVIEW

File: src/metabolic_ninja/helpers.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+def compute_chemical_linkage_strength(counts_a, counts_d):
+    """
+    Compute the strength of chemical linkage (SCL) as in [1]_.
+
+    Parameters
+    ----------
+    counts_a : dict
+        A mapping between atoms as characters and their counts as parsed from a
+        formula.
+    counts_d : dict
```
Comment (ChristianLieven) [inline src/metabolic_ninja/helpers.py:47]: Why did you choose counts_a and counts_d when we're comparing compound A vs B?

File: src/metabolic_ninja/helpers.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    for atom in atoms:
+        a_freq = counts_a.get(atom, 0)
+        d_freq = counts_d.get(atom, 0)
+        a_total += a_freq
+        d_total += d_freq
+        scl += min(a_freq, d_freq)
+    return scl / max(a_total, d_total)
+
+
+def identify_exotic_cofactors(pathway, solution, threshold=1E-07):
```
Comment (ChristianLieven) [inline src/metabolic_ninja/helpers.py]: Could you add tests for this function too? There is quite a lot of logic which means that tests for `compute_chemical_linkage_strength` and `count_atoms` don't cover much of this.

**Review by ChristianLieven — COMMENTED** [review]
Unfortunately, I didn't quite get all the way through. The idea of using SCL is smart, but I'm wondering if it holds up with co-factors that transfer carboxy groups or any other cofactor which doesn't only transfer electrons. The logic in `helpers` is only half tested, so perhaps there are some edge cases with a handful of cofactors you may want to test.

As I can see the logic in `tasks` also isn't unit tested. I'm not innately familiar with the metabolic-ninja yet but I'd wager that debugging anything will be made easier with a high coverage of unit tests. Concise tests certainly make code reviews easier as you can see the inputs and outputs at a quick glance :) 

In any case I hope that @kvikshaug could take over from here as I only got half-way through!

File: src/metabolic_ninja/tasks.py [hunk 3]
```diff
@@ -290,26 +298,64 @@ def heuristic_optimization(pathway, model):
...
+        The model of interest.
+
+    Returns
+    -------
+    list
+        The same list of results where each result has an added element
+        ``'exotic_cofactors'``.
+
+    """  # noqa: E501
+    # TODO: Get the tolerance from the solver and use it as the threshold.
```
Comment (kvikshaug) [inline src/metabolic_ninja/tasks.py]: One convention I've seen used is adding a name to comments like this, e.g., `TODO(Moritz Beber): ...`. The named person is not necessarily the one who wrote the comment, but someone who knows what it is about and could answer questions other devs might have. Perhaps that could be useful here?

File: src/metabolic_ninja/helpers.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                                                  count_atoms(target.formula))
+            scl = compute_chemical_linkage_strength(
+                count, target_count)
+            options.append((compound, scl))
+        # The maximum SCL is our candidate precursor.
+        options.sort(key=itemgetter(1))
+        logger.debug("Target: %r", target.id)
+        logger.debug("Precursor options: %r", options)
+        target, scl = options.pop()
+        logger.info("Picking %r (%r) with SCL = %.3G.", target.id,
```
Comment (kvikshaug) [inline src/metabolic_ninja/helpers.py]: Not a big deal, but I prefer that we use new-style formatting consistently over the performance gain.

File: src/metabolic_ninja/helpers.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    # not adapters nor exchanges.
+    rxn_queue = list(target.reactions & heterologous_reactions)
+    seen = set()
+    exotic_cofactors = set()
+    while len(rxn_queue) > 0:
+        rxn = rxn_queue.pop()
+        seen.add(rxn)
+        flux = 0.0 if abs(solution[rxn.id]) <= threshold else solution[rxn.id]
+        if flux > 0:
+            substrates = set(rxn.reactants)
```
Comment (kvikshaug) [inline src/metabolic_ninja/helpers.py:137]: Why are we naming these substrates instead of reactants?

File: src/metabolic_ninja/helpers.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+            # Generate the atom counts from the formulae.
+            count = atom_counts.setdefault(compound,
+                                           count_atoms(compound.formula))
+            target_count = atom_counts.setdefault(target,
+                                                  count_atoms(target.formula))
+            scl = compute_chemical_linkage_strength(
+                count, target_count)
+            options.append((compound, scl))
+        # The maximum SCL is our candidate precursor.
+        options.sort(key=itemgetter(1))
```
Comment (kvikshaug) [inline src/metabolic_ninja/helpers.py:165]: Suggestion: `max(options, key=itemgetter(1))` would perhaps describe the intention here even more clearly.

File: tests/test_unit/test_helpers.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    PathwayResult)
+
+import metabolic_ninja.helpers as helpers
+
+
+model_registry = {}
+pathway_registry = {}
+
+
+def register_with(registry):
```
Comment (kvikshaug) [inline tests/test_unit/test_helpers.py:31]: This registry pattern looks a bit odd to me, how does this work? I suppose it's got something to do with the `indirect` argument to `pytest.parametrize`?

**Review by kvikshaug — COMMENTED** [review]
Overall this looks very good! I mostly have questions about parts I don't understand, I am for example a little bit confused about the terminology (reactant vs substrate vs precursor). I trust that e.g. the chemical linkage strength algorithm correctly implements the cited paper.

**Review by kvikshaug — APPROVED** [review]
Had a chat with Moritz and got all the questions answered :+1:

File: src/metabolic_ninja/helpers.py (hunk 1 above)
Comment (Midnighter) [reply src/metabolic_ninja/helpers.py:47]: I have corrected the docstring. A and D are used in the paper.

File: src/metabolic_ninja/helpers.py (hunk 6 above)
Comment (Midnighter) [reply src/metabolic_ninja/helpers.py:165]: I decided to keep it in order to have the logging output sorted.
